### PR TITLE
Feature: Progress tracking and conditional termination from Python #206

### PR DESCRIPTION
**Pull request by mgevor**

### CODE REVIEW

File: include/usearch/index.hpp [hunk 1]
```diff
@@ -2721,8 +2731,13 @@ class index_gt {
         using slot_level_allocator_t = typename dynamic_allocator_traits_t::template rebind_alloc<slot_level_t>;
         buffer_gt<slot_level_t, slot_level_allocator_t> slots_and_levels(size());
 
+        // Progress status
+        std::atomic<bool> do_tasks{true};
```
Comment (ashvardanian) [inline include/usearch/index.hpp:2735]: It's better to store the atomic error in an `std::atomic<char const *>`, like in other palces.

File: include/usearch/index.hpp [hunk 2]
```diff
@@ -3411,6 +3441,9 @@ static join_result_t join(               //
     std::atomic<std::size_t> visited_members{0};
     std::atomic<char const*> atomic_error{nullptr};
 
+    // Progress status
+    std::atomic<bool> do_tasks{true};
```
Comment (ashvardanian) [inline include/usearch/index.hpp:3445]: No need for one more atomic variable.

File: include/usearch/index.hpp [hunk 3]
```diff
@@ -3431,7 +3464,10 @@ static join_result_t join(               //
                 passed_rounds = ++rounds;
                 total_rounds = passed_rounds + free_men.size();
             }
-            progress(passed_rounds, total_rounds);
+            if (thread_idx == 0)
+                do_tasks = progress(passed_rounds, total_rounds);
+            if (do_tasks.load())
```
Comment (ashvardanian) [inline include/usearch/index.hpp:3469]: Isn't it the opposite condition?

**Review by ashvardanian — CHANGES_REQUESTED** [review]

File: include/usearch/index.hpp (hunk 1 above)
Comment (mgevor) [reply include/usearch/index.hpp:2735]: This function doesn't return anything, so there is object to assign the error

**Comment by ashvardanian** [comment]
I believe the `index.py` file must be affected by these changes, but I can't see it among the changes.
I expect `tqdm` to receive progress reports from the C++ level. In that case - `batch_size` argument for `add` and `search` will become irrelevant at the Python level, and only `log` argument will remain. @mgevor, please update accordingly.

**Review by ashvardanian — APPROVED** [review]

**Pull request merged**